<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Introducción a Matplotlib y a las Gráficas de Líneas</font></h1>


## Introducción

El objetivo de estos laboratorios es darte una introducción a la visualización de datos con Python, tan concreto y conciso como sea posible.
Hablando de consistencia, debido a que no existe la *mejor* librería para visualizar datos en Python – hasta la fecha de creación de estos laboratorios – tendremos que mostrar distintas librerías y ensenar sus beneficios cuando discutamos nuevos conceptos sobre visualización. De esta forma esperamos que los estudiantes se familiaricen con librerías y conceptos para que puedan juzgar y decidir la mejor técnica y herramienta para visualizar dado el problema *y* la audiencia.

Asegurate de haber completado los requisitos previos para este curso, estos son [**Python Basics for Data Science**](https://www.edx.org/course/python-basics-for-data-science-2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01) and [**Analyzing Data with Python**](https://www.edx.org/course/data-analysis-with-python?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01).

**Observación**: La mayoría de las graficas y visualización serán generadas usando datos almacenados en DataFrames *pandas*. Por consiguiente, en este laboratorio, daremos un curso rápido de *pandas*. Sin embargo, si tienes interés en aprender mas sobre la librería *pandas*, una descripción detallada y explicaciones sobre como limpiar y procesar datos en un dataframe *pandas* están disponibles en nuestro curso [**Analyzing Data with Python**](https://www.edx.org/course/data-analysis-with-python?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01).

***


## Tabla of Contenido

<div class="alert alert-block alert-info" style="margin-top: 20px">

1.  [Exploración de Conjuntos de Datos con *pandas*](#0)<br>

1.1 [El Conjunto de Datos: Inmigración en Canadá desde 1980 a 2013](#2)<br>
1.2 [Conceptos básicos de *pandas*](#4) <br>
1.3 [Conceptos Intermedios de *pandas*: Indexado y Selección](#6) <br>
2\. [Visualizar Datos con Matplotlib](#8) <br>
2.1 [Matplotlib: Librería Estándar de Visualización para Python](#10) <br>
3\. [Gráficas de Líneas](#12)

</div>
<hr>


# Exploración de Conjuntos de Datos con *pandas* <a id="0"></a>

*pandas* es un kit de herramientas para el análisis de datos. De su \[sitio web] ([http://pandas.pydata.org/](http://pandas.pydata.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01)):

> *pandas* es un paquete de Python que ofrece estructuras de datos rápidas, flexibles y expresivas diseñadas para facilitar el trabajo con datos “relacionales” o “etiquetados”. Su objetivo es ser el bloque de construcción principal para realizar análisis practicas **del mundo real** en Python.

El curso tiene como base principal el uso de *pandas* para la exploración, análisis y visualización de datos. Te recomendamos invertir un poco de tu tiempo para familiarizarte con la API de referencia de *pandas*:[http://pandas.pydata.org/pandas-docs/stable/api.html](http://pandas.pydata.org/pandas-docs/stable/api.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01).


## El Conjunto de Datos: Inmigración en Canadá desde 1980 a 2013 <a id="2"></a>


Fuente de los Datos: [Flujos migratorios internacionales a y desde los países seleccionados – Revisión de 2015 ](http://www.un.org/en/development/desa/population/migration/data/empirical2/migrationflows.shtml?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01).

El conjunto de datos contiene información anual sobre los flujos de migración internacional recopilada por los países de destino. Los datos representan flujos de entrada y salida de acuerdo con el lugar de nacimiento, ciudadanía o lugar de residencia anterior/próxima tanto para extranjeros como nacionales.  La versión actual presenta datos provenientes de 45 paises.

En esta sesión, nos enfocaremos en los datos Migratorios de Canadá.

<img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DV0101EN/labs/coursera/Images/Mod1Fig1-Dataset.png" align="center" width=900>

Por cuestiones de simplicidad, los datos migratorios de Canadá han sido extraídos y cargados a uno de los servidores de IBM. Puedes obtener los datos desde [aquí](https://ibm.box.com/shared/static/lw190pt9zpy5bd1ptyg2aw15awomz9pu.xlsx?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01).

***


## Conceptos básicos de *pandas*<a id="4"></a>


Lo primero que haremos será importar dos módulos esenciales para el análisis de datos: *pandas* y *Numpy*.


In [11]:
import numpy as np  # muy útil para cálculos científicos con Python
import pandas as pd # Librería para estructar datos primarios
!pip install -U sqlalchemy==1.3.9
!pip install -U ibm_db_sa
!pip install -U ipython-sql
!pip install -U ibm-db
!pip install -U openpyxl
import ibm_db

     |████████████████████████████████| 242 kB 26.0 MB/s eta 0:00:01


Vamos a descargar e importar el conjunto principal de datos sobre Inmigración en Canadá utilizando el método *pandas* `read_excel()`. Normalmente, antes de hacerlo, necesitaremos descargar un modulo requerido por *pandas* para poder leer archivos de Excel. Este modulo es **xlrd**. Para tu comodidad, hemos preinstalado dicho modulo. De no ser así, tendrías que ejecutar la siguiente línea de código para instalar el modulo **xlrd**:

```
!conda install -c anaconda xlrd --yes
```


Ahora estamos listos para leer los datos.


In [8]:
df_can = pd.read_excel('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DV0101EN/labs/Data_Files/Canada.xlsx',
                       sheet_name='Canada by Citizenship',
                       skiprows=range(20),
                       skipfooter=2)

print ('Data read into a pandas dataframe!')

NameError: name 'pd' is not defined

Veamos las primeras 5 filas de nuestro conjunto de datos con la función `head()`


In [14]:
df_can.head()
# tip: puedes especificar el numero de filas que quieres ver de la siguiente manera: df_can.head(10)

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  ...  2004  2005  2006  \
0    Southern Asia  902  Developing regions    16  ...  2978  3436  3009   
1  Southern Europe  901   Developed regions     1  ...  1450  1223   856   
2  Northern Africa  902  Developing regions    80  ...  3616  3626  4807   
3        Polynesia  902  Developing regions     0  ...     0     0     1   
4  Southern Europe  901   Developed regions     0  ...     0     0     1   

   2007  2008  2009  2010  2011  2012  2013  
0  2652  2111  1746  1758  2203  2635  2004  
1   702   560   716   561   539   620   603  
2  3623  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0     0  
4     1     0     0     0     0     1     1  

[5 rows x 43 columns]

También podemos ver las 5 filas del final en nuestro conjunto de datos con la función `tail()`.


In [15]:
df_can.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

Cuando se analiza un conjunto de datos, es una buena idea empezar obteniendo información básica sobre tu dataframe. Podemos hacer esto con el método `info()`.


In [16]:
df_can.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 43 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Type      195 non-null    object
 1   Coverage  195 non-null    object
 2   OdName    195 non-null    object
 3   AREA      195 non-null    int64 
 4   AreaName  195 non-null    object
 5   REG       195 non-null    int64 
 6   RegName   195 non-null    object
 7   DEV       195 non-null    int64 
 8   DevName   195 non-null    object
 9   1980      195 non-null    int64 
 10  1981      195 non-null    int64 
 11  1982      195 non-null    int64 
 12  1983      195 non-null    int64 
 13  1984      195 non-null    int64 
 14  1985      195 non-null    int64 
 15  1986      195 non-null    int64 
 16  1987      195 non-null    int64 
 17  1988      195 non-null    int64 
 18  1989      195 non-null    int64 
 19  1990      195 non-null    int64 
 20  1991      195 non-null    int64 
 21  1992      195 no

Para obtener la lista de las cabeceras de las columnas podemos hacer uso del parámetro del dataframe `.columns`.


In [17]:
df_can.columns.values 

array(['Type', 'Coverage', 'OdName', 'AREA', 'AreaName', 'REG', 'RegName',
       'DEV', 'DevName', 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987,
       1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998,
       1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
       2010, 2011, 2012, 2013], dtype=object)

De manera similar para obtener una lista de los índices podemos usar el parámetro  `.index`.


In [18]:
df_can.index.values

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

Observación: El tipo de dato por defecto de los índices y columnas NO es lista (list).


In [19]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


Para obtener los índices y columnas como tipo lista (list) podemos usar el método `tolist()`


In [21]:
df_can.columns.tolist()
df_can.index.tolist()

print (type(df_can.columns.tolist()))
print (type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


Para ver las dimensiones del dataframe usamos el parámetro `.shape`.


In [22]:
# tamaño del dataframe (filas, columnas)
df_can.shape    

(195, 43)

Observación: Los tipos principales almacenados en objetos *pandas* son *float*, *int*, *bool*, *datetime64\[ns]* y *datetime64\[ns, tz] (in >= 0.17.0)*, *timedelta\[ns]*, *category (in >= 0.15.0)*, y *object*. Además estos tipos de datos posen tamaño, por ejemplo int64 and int32.


Limpiemos el conjunto de datos para retirar columnas innecesarias. Podemos usar el método de *pandas* `drop()` de la siguiente manera:


In [23]:
# En pandas axis=0 representa filas (por defecto) y axis=1 representa columnas.
df_can.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
df_can.head(2)

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1      Albania   Europe  Southern Europe   Developed regions     1     0   

   1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0    39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
1     0     0     0     0  ...  1450  1223   856   702   560   716   561   

   2011  2012  2013  
0  2203  2635  2004  
1   539   620   603  

[2 rows x 38 columns]

Renombremos las columnas para que tengan sentido. Podemos usar el método `rename()` pasándole un diccionario con nombres viejos y nuevos de la siguiente manera:


In [27]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.columns

Index([  'Country', 'Continent',    'Region',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013,     'Total'],
      dtype='object')

Vamos a añadir también una columna 'Total' para el numero total de inmigrantes por país a lo largo del periodo 1980 – 2013 de la siguiente manera:


In [28]:
df_can['Total'] = df_can.sum(axis=1)

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


Podemos revisar cuantos objetos vacíos (null) tenemos en el conjunto de datos de esta forma:


In [29]:
df_can.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

Para finalizar, veamos un resumen rápido de cada columna en nuestro dataframe usando el método `describe()`.


In [30]:
df_can.describe()

1980          1981          1982          1983          1984  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     508.394872    566.989744    534.723077    387.435897    376.497436   
std     1949.588546   2152.643752   1866.997511   1204.333597   1198.246371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     13.000000   
75%      251.500000    295.500000    275.000000    173.000000    181.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   195.000000   195.000000    195.000000    195.000000    195.000000   
mean    358.861538   441.271795    691.133333    714.389744    843.241026   
std    1079.309600  1225.576630   2109.205607   2443.606788   2555.048874   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.500000      0.500000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     34.000000     44.000000   
75%     197.000000   254.000000    434.000000    409.000000    508.500000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

       ...          2005          2006          2007          2008  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...   1320.292308   1266.958974   1191.820513   1246.394872   
std    ...   4425.957828   3926.717747   3443.542409   3694.573544   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     28.500000     25.000000     31.000000     31.000000   
50%    ...    210.000000    218.000000    198.000000    205.000000   
75%    ...    832.000000    842.000000    899.000000    934.500000   
max    ...  42584.000000  33848.000000  28742.000000  30037.000000   

               2009          2010          2011          2012          2013  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1275.733333   1420.287179   1262.533333   1313.958974   1320.702564   
std     3829.630424   4462.946328   4030.084313   4247.555161   4237.951988   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       36.000000     40.500000     37.500000     42.500000     45.000000   
50%      214.000000    211.000000    179.000000    233.000000    213.000000   
75%      888.000000    932.000000    772.000000    783.000000    796.000000   
max    29622.000000  38617.000000  36765.000000  34315.000000  34129.000000   

              Total  
count  1.950000e+02  
mean   9.860235e+04  
std    2.753565e+05  
min    3.000000e+00  
25%    2.856000e+03  
50%    1.505400e+04  
75%    6.671850e+04  
max    2.075712e+06  

[8 rows x 35 columns]

***

## Conceptos Intermedios de *pandas*: Indexado y Selección (slicing)<a id="6"></a>


### Seleccionar Columna

**Hay dos maneras de filtrar por el nombre de una columna:**

Método 1: Rápido y fácil, pero solo funcionará si el nombre de la columna NO tiene caracteres especiales o espacios.

```python
    df.column_name 
        (returns series)
```

Método 2: Es mas robusto y puede filtrar múltiples columnas.

```python
    df['column']  
        (returns series)
```

```python
    df[['column 1', 'column 2']] 
        (returns dataframe)
```

***


Ejemplo: Hagamos un filtro en la lista de países ('Country').


In [31]:
df_can.Country # regresa una serie

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Vamos a filtrar en la lista de países ('OdName') y en los datos para los años 1980 – 1985.


In [32]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # regresa un dataframe
# Observa que 'Country' es una cadena y los años son enteros.
# Con la finalidad de ser consistentes, convertiremos todos los nombres de las columnas a tipo cadenas mas adelante.

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Seleccionar Fila

Existen tres formas principales para seleccionar filas:

```python
    df.loc[label]        
        #filtra por la etiqueta del índice/columna
    df.iloc[index]       
        #filtra por la posición del índice/columna
```


Antes de proceder, observa que el índice por defecto del conjunto de datos es un rango numero desde 0 hasta 194. Esta dificulta hacer una consulta por un país en especifico. Por ejemplo para buscar información de Japón necesitamos su valor de índice correspondiente.

Esto puede resolverse con facilidad estableciendo la columna 'Country' como índice mediante el método `set_index()`.


In [33]:
df_can.set_index('Country', inplace=True)
# Tip: Lo opuesto de establecer (set) es reestablecer (reset). Para reestablecer el índice podemos usar df_can.reset_index().

In [34]:
df_can.head(3)

Continent           Region             DevName  1980  1981  1982  \
Country                                                                        
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   
Algeria        Africa  Northern Africa  Developing regions    80    67    71   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Country                              ...                                       
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Albania         0     0     0     1  ...  1223   856   702   560   716   561   
Algeria        69    63    44    69  ...  3626  4807  3623  4005  5393  4752   

             2011  2012  2013   Total  
Country                                
Afghanistan  2203  2635  2004  175917  
Albania       539   620   603   47097  
Algeria      4325  3774  4331  208317  

[3 rows x 38 columns]

In [36]:
# opcional: remueve el nombre del índice
df_can.index.name = None

Ejemplo: Veamos el numero de inmigrantes japoneses (fila 87) para los siguientes escenarios:
1\. El total de filas de datos (todas las columnas)
2\. Para el año 2013
3\. Para los años 1980 a 1985


In [37]:
# 1. El total de filas de datos (todas las columnas)
print(df_can.loc['Japan'])

# métodos alternativos
print(df_can.iloc[87])
print(df_can[df_can.index == 'Japan'].T.squeeze())

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [38]:
# 2. Para el año 2013
print(df_can.loc['Japan', 2013])

# método alternativo
print(df_can.iloc[87, 36]) # el año 2013 esta en la ultima columna con una posición de índice de 36

982
982


In [39]:
# 3. Para los años 1980 a 1985
print(df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]])
print(df_can.iloc[87, [3, 4, 5, 6, 7, 8]])

1980    701
1981    756
1982    598
1983    309
1984    246
1984    246
Name: Japan, dtype: object
1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object


Los nombres de las columnas que son enteros (como las de los años) puede provocar cierta confusión. Por ejemplo, cuando hacemos referencia al año 2013 uno pudiera confundirlo con la posición 2013 del índice.

Para evitar esta ambigüedad, vamos a convertir los nombres de las columnas a tipo cadena: '1980' a '2013'.


In [40]:
df_can.columns = list(map(str, df_can.columns))
# [print (type(x)) for x in df_can.columns.values] #<-- retirar el código comentado para revisar el tipo de dato de las cabeceras de las columnas

Debido a que convertimos los años a tipo cadena, declaremos una variable que nos permita fácilmente hacer una llamada al rango completo de años:


In [41]:
# de utilidad para graficar mas adelante
years = list(map(str, range(1980, 2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

### Filtrado basado en Criterios

Para filtrar un dataframe en base a una condición implemente debemos pasar la condición como vector boleano.

Por ejemplo, filtremos el dataframe para mostrar los datos acerca de los países asiáticos (AreaName = Asia).


In [42]:
# 1. crear las series de condiciones boleanas 
condition = df_can['Continent'] == 'Asia'
print(condition)

Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Viet Nam           True
Western Sahara    False
Yemen              True
Zambia            False
Zimbabwe          False
Name: Continent, Length: 195, dtype: bool


In [43]:
# 2. pasar al dataframe esta condición 
df_can[condition]

Continent              Region  \
Afghanistan                                         Asia       Southern Asia   
Armenia                                             Asia        Western Asia   
Azerbaijan                                          Asia        Western Asia   
Bahrain                                             Asia        Western Asia   
Bangladesh                                          Asia       Southern Asia   
Bhutan                                              Asia       Southern Asia   
Brunei Darussalam                                   Asia  South-Eastern Asia   
Cambodia                                            Asia  South-Eastern Asia   
China                                               Asia        Eastern Asia   
China, Hong Kong Special Administrative Region      Asia        Eastern Asia   
China, Macao Special Administrative Region          Asia        Eastern Asia   
Cyprus                                              Asia        Western Asia   
Democratic People's Republic of Korea               Asia        Eastern Asia   
Georgia                                             Asia        Western Asia   
India                                               Asia       Southern Asia   
Indonesia                                           Asia  South-Eastern Asia   
Iran (Islamic Republic of)                          Asia       Southern Asia   
Iraq                                                Asia        Western Asia   
Israel                                              Asia        Western Asia   
Japan                                               Asia        Eastern Asia   
Jordan                                              Asia        Western Asia   
Kazakhstan                                          Asia        Central Asia   
Kuwait                                              Asia        Western Asia   
Kyrgyzstan                                          Asia        Central Asia   
Lao People's Democratic Republic                    Asia  South-Eastern Asia   
Lebanon                                             Asia        Western Asia   
Malaysia                                            Asia  South-Eastern Asia   
Maldives                                            Asia       Southern Asia   
Mongolia                                            Asia        Eastern Asia   
Myanmar                                             Asia  South-Eastern Asia   
Nepal                                               Asia       Southern Asia   
Oman                                                Asia        Western Asia   
Pakistan                                            Asia       Southern Asia   
Philippines                                         Asia  South-Eastern Asia   
Qatar                                               Asia        Western Asia   
Republic of Korea                                   Asia        Eastern Asia   
Saudi Arabia                                        Asia        Western Asia   
Singapore                                           Asia  South-Eastern Asia   
Sri Lanka                                           Asia       Southern Asia   
State of Palestine                                  Asia        Western Asia   
Syrian Arab Republic                                Asia        Western Asia   
Tajikistan                                          Asia        Central Asia   
Thailand                                            Asia  South-Eastern Asia   
Turkey                                              Asia        Western Asia   
Turkmenistan                                        Asia        Central Asia   
United Arab Emirates                                Asia        Western Asia   
Uzbekistan                                          Asia        Central Asia   
Viet Nam                                            Asia  South-Eastern Asia   
Yemen                                               Asia        Western Asia   

                                              

In [44]:
# Podemos pasar múltiples condiciones en la misma línea
# filtremos por AreaNAme = Asia y RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# observación: cuando usamos los operadores 'and' y 'or', panda requiere el uso de '&' y '|' en lugar de 'and' y 'or'
# no olvides poner entre paréntesis las dos condiciones

Continent         Region             DevName  1980  \
Afghanistan                     Asia  Southern Asia  Developing regions    16   
Bangladesh                      Asia  Southern Asia  Developing regions    83   
Bhutan                          Asia  Southern Asia  Developing regions     0   
India                           Asia  Southern Asia  Developing regions  8880   
Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions  1172   
Maldives                        Asia  Southern Asia  Developing regions     0   
Nepal                           Asia  Southern Asia  Developing regions     1   
Pakistan                        Asia  Southern Asia  Developing regions   978   
Sri Lanka                       Asia  Southern Asia  Developing regions   185   

                            1981  1982  1983  1984  1985  1986  ...   2005  \
Afghanistan                   39    39    47    71   340   496  ...   3436   
Bangladesh                    84    86    81    98    92   486  ...   4171   
Bhutan                         0     0     0     1     0     0  ...      5   
India                       8670  8147  7338  5704  4211  7150  ...  36210   
Iran (Islamic Republic of)  1429  1822  1592  1977  1648  1794  ...   5837   
Maldives                       0     0     1     0     0     0  ...      0   
Nepal                          1     6     1     2     4    13  ...    607   
Pakistan                     972  1201   900   668   514   691  ...  14314   
Sri Lanka                    371   290   197  1086   845  1838  ...   4930   

                             2006   2007   2008   2009   2010   2011   2012  \
Afghanistan                  3009   2652   2111   1746   1758   2203   2635   
Bangladesh                   4014   2897   2939   2104   4721   2694   2640   
Bhutan                         10      7     36    865   1464   1879   1075   
India                       33848  28742  28261  29456  34235  27509  30933   
Iran (Islamic Republic of)   7480   6974   6475   6580   7477   7479   7534   
Maldives                        0      2      1      7      4      3      1   
Nepal                         540    511    581    561   1392   1129   1185   
Pakistan                    13127  10124   8994   7217   6811   7468  11227   
Sri Lanka                    4714   4123   4756   4547   4422   3309   3338   

                             2013    Total  
Afghanistan                  2004   175917  
Bangladesh                   3789   196704  
Bhutan                        487    17628  
India                       33087  2075712  
Iran (Islamic Republic of)  11291   527769  
Maldives                        1       90  
Nepal                        1308    30666  
Pakistan                    12603   724800  
Sri Lanka                    2394   445074  

[9 rows x 38 columns]

Antes de continuar: revisemos los cambios que hemos hecho en el dataframe.


In [7]:
print('data dimensions:', df_can.shape)
print(df_can.columns)
df_can.head(2)

NameError: name 'df_can' is not defined

***

# Visualizar Datos con Matplotlib<a id="8"></a>


## Matplotlib: Librería Estándar de Visualización para Python<a id="10"></a>

La primera librería para graficado que exploraremos en este curso será [Matplotlib](http://matplotlib.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01). Como se menciona en su sitio web:

> Matplotlib es una librería para graficar de Python 2D y que produce imágenes y publicaciones de calidad en una variedad de formatos y entornos interactivos a través de varias plataformas. Matplotlib puede usarse con scripts de Python, la Shell de Python y IPython, cuadernos de jupyter, servidores de aplicaciones de web y cuatro kits de herramientas para interfaces graficas de usuario.

Si quieres crear visualizaciones impactantes con Python, Matplotlib es una herramienta esencial que tienes a tu disposición.


### Matplotlib.Pyplot

Uno de los aspectos centrales de Matplotlib es `matplotlib.pyplot`. Es la capa de scripting de Matplotlib que hemos estudiado en detalle en los videos acerca de Matplotlib. Recuerda que es una colección de funciones de tipo comando que hacen que Matplotlib trabaje junto a MATLAB. Cada función `pyplot` provoca algún cambio en una imagen, p.ej. crear una figura, crear un área de dibujo en una figura, graficar algunas líneas, decorar la grafica con etiquetas, etc. En este laboratorio trabajaremos con la capa de scripting con el fin de aprender a generar graficas de líneas. En próximos laboratorios trabajaremos con la capa de artista también para experimentar de primera mano sus diferencias con la capa de scripting.


Empecemos importando `Matplotlib` y `Matplotlib.pyplot` de la siguiente manera:


In [6]:
# se usara el backend en línea
%matplotlib inline 

import matplotlib as mpl
import matplotlib.pyplot as plt

\*opcional: revisar si Matplotlib esta cargado.


In [4]:
print ('Matplotlib version: ', mpl.__version__) # >= 2.0.0

Matplotlib version:  3.4.3


\*opcional: aplicar un estilo a Matplotlib


In [5]:
print(plt.style.available)
mpl.style.use(['ggplot']) # opcional: para el estilo tipo ggplot

['Solarize_Light2', '_classic_test_patch', 'bmh', 'classic', 'dark_background', 'fast', 'fivethirtyeight', 'ggplot', 'grayscale', 'seaborn', 'seaborn-bright', 'seaborn-colorblind', 'seaborn-dark', 'seaborn-dark-palette', 'seaborn-darkgrid', 'seaborn-deep', 'seaborn-muted', 'seaborn-notebook', 'seaborn-paper', 'seaborn-pastel', 'seaborn-poster', 'seaborn-talk', 'seaborn-ticks', 'seaborn-white', 'seaborn-whitegrid', 'tableau-colorblind10']


### Graficar con *pandas*

Afortunadamente pandas tiene una implementación pre cargada de Matplotlib que podemos usar. Graficar en *pandas* es tan simple como añadir un método `.plot()` a una serie o dataframe.

Documentación:

*   [Graficación con Series](http://pandas.pydata.org/pandas-docs/stable/api.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01#plotting)<br>
*   [Graficación con Dataframes](http://pandas.pydata.org/pandas-docs/stable/api.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01#api-dataframe-plotting)


# Gráficas de Líneas (Series/Dataframe) <a id="12"></a>


**¿Qué es una grafica de línea y para que se usa?**

Una grafica de línea muestra información como una seria de puntos de datos llamados 'marcadores' conectados por segmentos de líneas rectas. Es un tipo de grafica común en muchos campos.
Utiliza graficas de líneas cuando tengas conjuntos de datos continuos. Estas graficas son ideales para visualización que muestran una tendencia de los datos a lo largo de un periodo de tiempo.


**Empecemos con un caso de estudio:**

En 2010 Haití sufrió un terremoto catastrófico de 7.0 grados de magnitud. Esto causo devastación y perdidas de vida en grandes áreas y afectando alrededor de tres millones de personas. Como parte de los esfuerzos humanitarios por parte de Canadá, el gobierno acepto refugiados provenientes de Haití. Podemos visualizar rápidamente este esfuerzo con el uso de una grafica de `Línea:`

**Pregunta:** Dibuja una Grafica de Línea de la inmigración proveniente de Haití usando `df.plot()`.


Primero, extraeremos la serie de datos de Haití.


In [1]:
haiti = df_can.loc['Haiti', years] # ejecución en los años 1980 a 2013 para excluir la columna 'total'
haiti.head()

NameError: name 'df_can' is not defined

Después, dibujaremos una grafica de línea anexando `.plot()` al dataframe `haiti`


In [2]:
haiti.plot()

NameError: name 'haiti' is not defined

*pandas* completó automáticamente el eje x con los valores de los índices (años) y el eje y con los valores de las columnas (población). Sin embargo, observe que los años no se mostraron porque son del tipo cadena. Por lo tanto, modifiquemos el tipo de los valores de índice a entero para su diagramación.

Además, etiquetemos el eje x y el eje y con `plt.title()`, `plt.ylabel()` y `plt.xlabel()` de la manera siguiente:


In [9]:
haiti.index = haiti.index.map(int) # cambiemos los valores de índice de Haiti al tipo entero para la graficación
haiti.plot(kind='line')

plt.title('Immigration from Haiti')
plt.ylabel('Number of immigrants')
plt.xlabel('Years')

plt.show() # esta línea es necesaria para mostrar las actualizaciones en la imagen

NameError: name 'haiti' is not defined

Podemos observar con claridad la mayor cantidad de inmigrantes desde Haití a partir de 2010, año en que Canadá aumentó sus esfuerzos para aceptar refugiados haitianos. Anotemos este pico en el diagrama con el método `plt.text()`.


In [ ]:
haiti.plot(kind='line')

plt.title('Immigration from Haiti')
plt.ylabel('Number of Immigrants')
plt.xlabel('Years')

# anotar el terremoto de 2010.
# sintaxis: plt.text(x, y, label)
plt.text(2000, 6000, '2010 Earthquake') # vea la nota abajo

plt.show() 

Con unas pocas líneas de código, pudo identificar rápidamente el aumento en inmigración y visualizarlo.

Una nota rápida sobre los valores de x y de y en `plt.text(x, y, label)`:

```
 Debido a que el eje x (años) es del tipo 'enteros', especificamos x como un año. El eje y (cantidad de inmigrantes) es del tipo 'enteros'; por lo tanto, podemos especificar el valor de y = 6000.
```

```python
    plt.text(2000, 6000, '2010 Earthquake') # años almacenados como tipo entero
```

```
Si los años se hubieran almacenado como tipo 'cadena', deberíamos especificar x como la posición del índice para el año. P. ej., el índice 20.º es el año 2000, debido a que se trata del año 20.º a partir del año base de 1980.
```

```python
    plt.text(20, 6000, '2010 Earthquake') # años almacenados como tipo enteros
```

```
Trataremos los métodos de anotación avanzados en módulos posteriores.
```


Podemos agregar fácilmente más países al gráfico de líneas para llevar a cabo comparaciones significativas en la inmigración desde distintos países.

**Pregunta:** Comparemos la cantidad de inmigrantes desde India y China entre 1980 a 2013.


Paso 1: Obtenga el conjunto de datos para China e India, y muestre el DataFrame.


In [10]:
### escribe aquí tu respuesta
df_CI = df_can.loc[['India', 'China'], years]
df_CI.head()


NameError: name 'df_can' is not defined

Haz doble clic **aquí** para ver la solución.

<!-- The correct answer is:
df_CI = df_can.loc[['India', 'China'], years]
df_CI.head()
-->


Paso 2: Dibuje el gráfico. Especificaremos explícitamente el gráfico de líneas al ejecutar el parámetro `kind` en `plot()`.


In [ ]:
### escribe aquí tu respuesta
df_CI.plot(kind='line')


Haz doble clic **aquí** para ver la solución.

<!-- The correct answer is:
df_CI.plot(kind='line')
-->


¡Eso no se ve correcto!

Recuerda que pandas dibuja los índices en el eje x y las columnas como líneas individuales en el eje y. Debido a que `df_CI` es un DataFrame con `country` como índice y years como columnas, debemos primero transponer el DataFrame con el método `transpose()` a fin de intercambiar la fila y las columnas.


In [ ]:
df_CI = df_CI.transpose()
df_CI.head()

*pandas* creará automáticamente un gráfico con los dos países. Grafique el nuevo DataFrame transpuesto. Asegúrese de agregar un título al diagrama y de etiquetar los ejes.


In [ ]:
### escribe aquí tu respuesta
df_CI.index = df_CI.index.map(int) # let's change the index values of df_CI to type integer for plotting
df_CI.plot(kind='line')

plt.title('Immigrants from China and India')
plt.ylabel('Number of Immigrants')
plt.xlabel('Years')

plt.show()




Haz doble clic **aquí** para ver la solución.

<!-- The correct answer is:
df_CI.index = df_CI.index.map(int) # let's change the index values of df_CI to type integer for plotting
df_CI.plot(kind='line')
-->

<!--
plt.title('Immigrants from China and India')
plt.ylabel('Number of Immigrants')
plt.xlabel('Years')
-->

<!--
plt.show()
--> 


A partir de la grafica anterior, podemos observar que China e India tienen tendencias inmigratorias muy similares a lo largo de los años.


*Observación*: ¿Por qué no tuvimos que transponer el DataFrame de Haití antes de diagramarlo (como hicimos con df_CI)? did for df_CI)?

Porque `haiti` es una serie, en lugar de un DataFrame, y los años son sus índices, como se muestra debajo.

```python
print(type(haiti))
print(haiti.head(5))
```

> class 'pandas.core.series.Series' <br>
> 1980    1666 <br>
> 1981    3692 <br>
> 1982    3498 <br>
> 1983    2860 <br>
> 1984    1418 <br>
> Name: Haiti, dtype: int64 <br>


El gráfico de líneas es una herramienta práctica para mostrar varias variables dependientes contra una variable independiente. Sin embargo, no se recomiendan más de 5 a 10 líneas por gráfico. Si hay más líneas, se hace difícil interpretar el gráfico.


**Pregunta:** Compara la tendencia de los 5 países principales que aportaron la mayor cantidad de inmigrantes a Canadá.


In [ ]:
### escribe aquí tu respuesta





Haz doble clic **aquí** para ver la solución.

<!-- The correct answer is:
\\\\ # Step 1: Get the dataset. Recall that we created a Total column that calculates the cumulative immigration by country. \\ We will sort on this column to get our top 5 countries using pandas sort_values() method.
\\\\ inplace = True paramemter saves the changes to the original df_can dataframe
df_can.sort_values(by='Total', ascending=False, axis=0, inplace=True)
-->

<!--
# get the top 5 entries
df_top5 = df_can.head(5)
-->

<!--
# transpose the dataframe
df_top5 = df_top5[years].transpose() 
-->

<!--
print(df_top5)
-->

<!--
\\\\ # Step 2: Plot the dataframe. To make the plot more readeable, we will change the size using the `figsize` parameter.
df_top5.index = df_top5.index.map(int) # let's change the index values of df_top5 to type integer for plotting
df_top5.plot(kind='line', figsize=(14, 8)) # pass a tuple (x, y) size
-->

<!--
plt.title('Immigration Trend of Top 5 Countries')
plt.ylabel('Number of Immigrants')
plt.xlabel('Years')
-->

<!--
plt.show()
-->


### Otros gráficos

¡Felicidades! Aprendiste a dominar datos con Python y a crear un gráfico de líneas con Matplotlib. Hay muchos más estilos de graficación disponibles, además del gráfico de líneas predeterminado. Puede acceder a todos ellos al ejecutar la palabra clave `kind` en `plot()`. La lista completa de diagramas disponibles es la siguiente:

*   `bar` para diagramas de barras verticales
*   `barh` para diagramas de barras horizontales
*   `hist` para histograma
*   `box` para diagrama de caja
*   `kde` o `density` para diagramas de densidad
*   `area` para diagramas de área
*   `pie` para diagramas circulares (de pastel)
*   `scatter` para diagramas de dispersión
*   `hexbin` para diagramas con agrupamiento hexagonal


### ¡Gracias por completar este laboratorio!

Este cuaderno fue creado originalmente por [Jay Rajasekharan](https://www.linkedin.com/in/jayrajasekharan?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01) con aportes de [Ehsan M. Kermani](https://www.linkedin.com/in/ehsanmkermani?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01) y [Slobodan Markovic](https://www.linkedin.com/in/slobodan-markovic?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01).

Recientemente, esta libreta fue modificada por [Alex Aklson](https://www.linkedin.com/in/aklson/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01). Espero que esta sesión de laboratorio le haya resultado interesante. ¡Si tiene alguna pregunta, comuníquese conmigo!


Este cuaderno es parte del curso en **edX** llamado *The Data Science Method*. Si tienes acceso a este documento desde fuera del curso, puedes tomarlo en línea haciendo clic [aquí](https://cocl.us/DS0103EN_edX_LAB4).


<hr>

Copyright © 2019 [Cognitive Class](https://cognitiveclass.ai/?utm_medium=dswb&utm_source=bducopyrightlink&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ESedX20297760-2021-01-01).
